### ISSUE TITLE
How about using inverse KL-divergence?

### ISSUE THREAD

**Issue by metalwhale** [jxiw/MambaInLlama#12]
Hello,
First of all, thank you so much for sharing your work; I really appreciate it!
I'm not very strong in math, so please take this with a grain of salt

As I see, your [distillation phase](https://github.com/jxiw/MambaInLlama/blob/03047eb/train_mamba/train_hybrid.py#L270-L271) is using [KL divergence as the loss](https://pytorch.org/docs/stable/generated/torch.nn.KLDivLoss.html#torch.nn.KLDivLoss) (with `true` = teacher (target), `pred` = student):
```
L(pred, true) = true * log(true / pred)
```

In [MiniLLM paper](https://arxiv.org/pdf/2306.08543), Gu et al. learned that using *reverse* KL divergence helps the student model accurately focus on the major parts of the target distribution, which is crucial for ensuring the correctness of text generation.
```
L(true, pred) = pred * log(pred / true)
```

How do you think?

**Comment by jxiw** [comment]
Thank you for sharing this. I believe it’s a wonderful idea. While I haven't yet had the opportunity to fully explore the advanced distillation approach, I am quite confident that by incorporating this method, we could significantly improve the model’s performance. At this stage, the distillation process is not yet fully optimized, so there is plenty of room for improvement.

**Comment by metalwhale** [comment]
@jxiw 
Thank you!
I love the idea of Mamba and believe it has great potential for exploration.
Always support your amazing work, and please send my thanks to your colleagues as well!

Looking forward to your next paper.
Issue closed with faith!
